benchmark_test.go

```diff
@@ -12,86 +12,92 @@ import (
+func (*BenchmarkLRUSuite) BenchmarkAddAndEvictInt0000020(c *gc.C) {
 	benchAddAndEvictInt(c, 20)
 }
 
-func (*BenchmarkSuite) BenchmarkAddAndEvictInt0000050(c *gc.C) {
+func (*BenchmarkLRUSuite) BenchmarkAddAndEvictInt0000050(c *gc.C) {
 	benchAddAndEvictInt(c, 50)
 }
 
-func (*BenchmarkSuite) BenchmarkAddAndEvictInt0000100(c *gc.C) {
+func (*BenchmarkLRUSuite) BenchmarkAddAndEvictInt0000100(c *gc.C) {
 	benchAddAndEvictInt(c, 100)
 }
 
-func (*BenchmarkSuite) BenchmarkAddAndEvictInt0000200(c *gc.C) {
+func (*BenchmarkLRUSuite) BenchmarkAddAndEvictInt0000200(c *gc.C) {
 	benchAddAndEvictInt(c, 200)
 }
 
-func (*BenchmarkSuite) BenchmarkAddAndEvictInt0000500(c *gc.C) {
+func (*BenchmarkLRUSuite) BenchmarkAddAndEvictInt0000500(c *gc.C) {
 	benchAddAndEvictInt(c, 500)
 }
 
-func (*BenchmarkSuite) BenchmarkAddAndEvictInt0001000(c *gc.C) {
+func (*BenchmarkLRUSuite) BenchmarkAddAndEvictInt0001000(c *gc.C) {
 	benchAddAndEvictInt(c, 1000)
 }
 
-func (*BenchmarkSuite) BenchmarkAddAndEvictInt0002000(c *gc.C) {
+func (*BenchmarkLRUSuite) BenchmarkAddAndEvictInt0002000(c *gc.C) {
 	benchAddAndEvictInt(c, 2000)
 }
 
-func (*BenchmarkSuite) BenchmarkAddAndEvictInt0005000(c *gc.C) {
+func (*BenchmarkLRUSuite) BenchmarkAddAndEvictInt0005000(c *gc.C) {
 	benchAddAndEvictInt(c, 5000)
 }
 
-func (*BenchmarkSuite) BenchmarkAddAndEvictInt0010000(c *gc.C) {
+func (*BenchmarkLRUSuite) BenchmarkAddAndEvictInt0010000(c *gc.C) {
 	benchAddAndEvictInt(c, 10000)
 }
 
-func (*BenchmarkSuite) BenchmarkAddAndEvictInt0020000(c *gc.C) {
+func (*BenchmarkLRUSuite) BenchmarkAddAndEvictInt0020000(c *gc.C) {
 	benchAddAndEvictInt(c, 20000)
 }
 
-func (*BenchmarkSuite) BenchmarkAddAndEvictInt0050000(c *gc.C) {
+func (*BenchmarkLRUSuite) BenchmarkAddAndEvictInt0050000(c *gc.C) {
 	benchAddAndEvictInt(c, 50000)
 }
 
-func (*BenchmarkSuite) BenchmarkAddAndEvictInt0100000(c *gc.C) {
+func (*BenchmarkLRUSuite) BenchmarkAddAndEvictInt0100000(c *gc.C) {
 	benchAddAndEvictInt(c, 100000)
 }
 
-func (*BenchmarkSuite) BenchmarkAddAndEvictInt0200000(c *gc.C) {
+func (*BenchmarkLRUSuite) BenchmarkAddAndEvictInt0200000(c *gc.C) {
 	benchAddAndEvictInt(c, 200000)
 }
 
-func (*BenchmarkSuite) BenchmarkAddAndEvictInt0500000(c *gc.C) {
+func (*BenchmarkLRUSuite) BenchmarkAddAndEvictInt0500000(c *gc.C) {
 	benchAddAndEvictInt(c, 500000)
 }
 
-func (*BenchmarkSuite) BenchmarkAddAndEvictInt1000000(c *gc.C) {
+func (*BenchmarkLRUSuite) BenchmarkAddAndEvictInt1000000(c *gc.C) {
 	benchAddAndEvictInt(c, 1000000)
 }
 
-func (*BenchmarkSuite) BenchmarkAddAndEvictInt2000000(c *gc.C) {
+func (*BenchmarkLRUSuite) BenchmarkAddAndEvictInt2000000(c *gc.C) {
 	benchAddAndEvictInt(c, 2000000)
 }
 
-func (*BenchmarkSuite) BenchmarkIntMemSize(c *gc.C) {
+func (*BenchmarkLRUSuite) BenchmarkIntMemSize(c *gc.C) {
+	keys := make([]int, c.N)
+	for i := 0; i < c.N; i++ {
+		keys[i] = i + 1e7
+	}
+	c.ResetTimer()
+	rand.Shuffle(c.N, func(i, j int) { keys[j], keys[i] = keys[i], keys[j] })
```

Comment:
Are we initializing the rand seed anywhere?

Reply:
I don't know that we are, though the key for this is more "don't add them as 1, 2, 3, 4, 5" than that it needs to be a different random each time.
we can put an 'init() { rand.Seed(time.Now().UnixNano()) }' if you like

Reply:
I would also be interested in seeing some performance numbers when the keys are a bit more randomized (e.g `		keys[i] = int(1e7 + rand.Int63n(1e7-1))`).

Even though the keys are shuffled, they are still linear; depending on how the underlying hashmap is implemented (linear probing comes to mind) you _may_ be getting more favorable performance from your inserts.

Reply:
fair enough. I think if we wanted to do it proper justice, we would use a benchmark that used non-unique keys. Something that would have a wide distribution and some keys more common than other keys, etc.
Most of these tests have essentially 0 hit rate, because all the keys are unique. So it does test churn of the algorithm, but not matching real-world use cases.